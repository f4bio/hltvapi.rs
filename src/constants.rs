pub const CREATE_DB_POOL_ERROR: &str = "Failed to create db pool";
pub const GET_DB_CONNECTION_ERROR: &str = "Failed to get db connection";
pub const BASE_URL: &str = "https://www.hltv.org";
pub const FILE_STORAGE_LOCATION: &str = "./static/";
pub const NEWS_URL: &str = "https://www.hltv.org/";
pub const NEWS_FEED_FILENAME: &str = "news.feed.xml";
pub const FIXTURES_URL_TOP_TIER: &str = "https://www.hltv.org/matches?predefinedFilter=top_tier";
pub const FIXTURES_URL: &str = "https://www.hltv.org/matches";
pub const FIXTURES_CALENDAR_FILENAME: &str = "fixtures.calendar.ical";
