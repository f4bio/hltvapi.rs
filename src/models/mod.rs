pub mod fixture;
pub mod news;
pub mod version;
