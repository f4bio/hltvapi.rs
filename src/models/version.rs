use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct Version {
  pub version: String,
}
