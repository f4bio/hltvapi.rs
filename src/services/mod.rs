use crate::errors::ApiError;

pub async fn do_fetch(url: &str) -> Result<String, ApiError> {
  // let proxy = reqwest::Proxy::http("https://my.prox").unwrap();
  let client = reqwest::Client::builder()
    // .proxy(proxy)
    .build()
    .unwrap();

  let resp = client.get(url).send().await;
  if resp.is_err() {
    let error_msg = format!(
      "received error while fetching news (code='{}')...",
      resp.as_ref().unwrap().status()
    );
    Err(ApiError::new(100, error_msg))
  } else {
    Ok(resp.unwrap().text().await.unwrap())
  }
}

pub mod fixtures;
pub mod news;
