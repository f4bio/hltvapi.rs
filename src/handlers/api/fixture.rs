use std::path::Path;

use actix_files::NamedFile;
use actix_web::http::header::{ContentDisposition, DispositionParam, DispositionType};
use actix_web::{get, web, HttpResponse};

use crate::constants::FIXTURES_CALENDAR_FILENAME;
use crate::errors::ApiError;
use crate::handlers::Limiter;
use crate::models::fixture::{Fixture, PublicFixture};

/// get all fixtures
#[get("/fixture/list")]
pub async fn find_all(params: web::Query<Limiter>) -> Result<HttpResponse, ApiError> {
  let inner_params = params.into_inner();
  let fixtures: Vec<PublicFixture> =
    Fixture::filtered_public(inner_params.limit, inner_params.top_tier)?;
  Ok(
    HttpResponse::Ok()
      .content_type(mime::APPLICATION_JSON)
      .json(fixtures),
  )
}

/// get news as ics calendar file
#[get("/fixture/calendar.ics")]
pub async fn calendar_ics() -> Result<NamedFile, ApiError> {
  let path = Path::new("static").join(FIXTURES_CALENDAR_FILENAME);
  let file = NamedFile::open(path).unwrap();
  Ok(
    file
      .set_content_disposition(ContentDisposition {
        disposition: DispositionType::Inline,
        parameters: vec![DispositionParam::Name(String::from(
          FIXTURES_CALENDAR_FILENAME,
        ))],
      })
      .set_content_type("text/calendar".parse().unwrap()),
  )
}

/// extract path info from "/fixture/{hashid}" url
/// {hashid} - deserializes to hashid
#[get("/fixture/{hash}")]
pub async fn find(hash: web::Path<String>) -> Result<HttpResponse, ApiError> {
  let fixtures = Fixture::find_public(hash.into_inner())?;
  Ok(
    HttpResponse::Ok()
      .content_type(mime::APPLICATION_JSON)
      .json(fixtures),
  )
}
