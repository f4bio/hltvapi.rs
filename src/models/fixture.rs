use chrono::{NaiveDateTime, Utc};
use diesel::prelude::*;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::database;
use crate::errors::ApiError;
use crate::schema::fixtures;

/// Representation of fixture
///
/// Results from data store are represented as vector of `Fixture` structs.
/// contents of the string. It has a close relationship with its borrowed
/// counterpart, the primitive [`str`].
///
/// # Examples
///
/// You can create a `String` from [a literal string][`str`] with [`String::from`]:
///
/// [`String::from`]: From::from
///
/// ```
/// let hello = String::from("Hello, world!");
/// ```
///
/// You can append a [`char`] to a `String` with the [`push`] method, and
/// append a [`&str`] with the [`push_str`] method:
#[derive(Serialize, Deserialize, AsChangeset, Queryable)]
#[table_name = "fixtures"]
pub struct Fixture {
  pub id: i32,
  pub name: String,
  pub link: String,
  pub team1: String,
  pub team2: String,
  pub start_time: NaiveDateTime,
  pub rating: i32,
  pub meta: String,
  pub analytics: String,
  pub top_tier: bool,
  pub hash: String,
  pub created_at: NaiveDateTime,
  pub updated_at: Option<NaiveDateTime>,
}

/// use new to send object to data store (sqlite)
#[derive(Deserialize, Serialize, AsChangeset, Insertable)]
#[table_name = "fixtures"]
pub struct NewFixture {
  pub name: String,
  pub link: String,
  pub team1: String,
  pub team2: String,
  pub start_time: NaiveDateTime,
  pub rating: i32,
  pub meta: String,
  pub analytics: String,
  pub top_tier: bool,
  pub hash: String,
}

/// use new to send object to data store (sqlite)
#[derive(Deserialize, Serialize)]
pub struct PublicFixture {
  pub name: String,
  pub link: String,
  pub team1: String,
  pub team2: String,
  pub start_time: NaiveDateTime,
  pub rating: i32,
  pub meta: String,
  pub analytics: String,
  pub top_tier: bool,
  pub hash: String,
  pub created_at: NaiveDateTime,
  pub updated_at: Option<NaiveDateTime>,
}

/// use dto to send data between modules
#[derive(Deserialize, Serialize)]
pub struct FixtureDTO {
  pub id: Option<i32>,
  pub name: Option<String>,
  pub link: Option<String>,
  pub team1: Option<String>,
  pub team2: Option<String>,
  pub start_time: Option<NaiveDateTime>,
  pub rating: Option<i32>,
  pub meta: Option<String>,
  pub analytics: Option<String>,
  pub top_tier: Option<bool>,
  pub hash: Option<String>,
  pub created_at: Option<NaiveDateTime>,
  pub updated_at: Option<NaiveDateTime>,
}

impl Fixture {
  pub fn all() -> Result<Vec<Self>, ApiError> {
    use crate::schema::fixtures::dsl::*;
    let conn = database::connection()?;
    let res: Vec<Fixture> = fixtures.load::<Fixture>(&conn)?;
    Ok(res)
  }

  pub fn all_public() -> Result<Vec<PublicFixture>, ApiError> {
    use crate::schema::fixtures::dsl::*;
    let conn = database::connection()?;
    let db_result: Vec<Fixture> = fixtures.load::<Fixture>(&conn)?;
    let res: Vec<PublicFixture> = db_result.into_iter().map(PublicFixture::from).collect();
    Ok(res)
  }

  pub fn filtered_public(
    limit_to: Option<i64>,
    is_top_tier: Option<bool>,
  ) -> Result<Vec<PublicFixture>, ApiError> {
    use crate::schema::fixtures::dsl::*;
    let conn = database::connection()?;
    let limit = limit_to.unwrap_or(10);
    let filter_top_tier = is_top_tier.unwrap_or(true);
    let db_result: Vec<Fixture> = fixtures
      .limit(limit)
      .filter(top_tier.eq(filter_top_tier))
      .load::<Fixture>(&conn)?;
    let res: Vec<PublicFixture> = db_result.into_iter().map(PublicFixture::from).collect();
    Ok(res)
  }

  pub fn get(id_to_get: i32) -> Result<Self, ApiError> {
    use crate::schema::fixtures::dsl::*;
    let conn = database::connection()?;
    let res = fixtures.filter(id.eq(id_to_get)).first(&conn)?;
    Ok(res)
  }

  pub fn find(hash_to_find: String) -> Result<Self, ApiError> {
    use crate::schema::fixtures::dsl::*;
    let conn = database::connection()?;
    let res = fixtures.filter(hash.eq(hash_to_find)).first(&conn)?;
    Ok(res)
  }

  pub fn find_public(hash_to_find: String) -> Result<PublicFixture, ApiError> {
    use crate::schema::fixtures::dsl::*;
    let conn = database::connection()?;
    let db_result: Fixture = fixtures.filter(hash.eq(hash_to_find)).first(&conn)?;
    let res: PublicFixture = PublicFixture::from(db_result);
    Ok(res)
  }

  /// create many new fixtures
  ///
  /// # Arguments
  ///
  /// * `incoming`: vector/array of new fixtures to create
  ///
  /// returns: Result<usize, ApiError>
  ///
  /// # Examples
  ///
  /// ```
  /// let mut fixtures: Vec<FixtureDTO> = Vec::new();
  /// fixtures.push(Fixture {/*...*/});
  /// fixtures.push(Fixture {/*...*/});
  ///
  /// let inserted_count = Fixture::create_many(fixtures);
  /// assert!(inserted_count,
  /// ```
  pub fn create(new_fixture: FixtureDTO) -> Result<Self, ApiError> {
    use crate::schema::fixtures::dsl::*;
    let conn = database::connection()?;
    // insert passed values:
    diesel::insert_or_ignore_into(fixtures)
      .values(NewFixture::from(new_fixture))
      .execute(&conn)?;
    // query and return first (most recent updated) item
    let res: Fixture = fixtures.first(&conn)?;
    Ok(res)
  }

  /// create many new fixtures.
  /// inserts are done by looping over the provided vector and inserting them one by one.
  /// sqlite doesnt support bulk inserts.
  ///
  /// # Arguments
  ///
  /// * `incoming`: vector/array of new fixtures to create
  ///
  /// returns: Result<usize, ApiError>
  ///
  /// # Examples
  ///
  /// ```
  /// let mut fixtures: Vec<FixtureDTO> = Vec::new();
  /// fixtures.push(Fixture {/*...*/});
  /// fixtures.push(Fixture {/*...*/});
  ///
  /// let inserted_count = Fixture::create_many(fixtures);
  /// assert!(inserted_count,
  /// ```
  pub fn create_many(incoming: Vec<FixtureDTO>) -> Result<usize, ApiError> {
    use crate::schema::fixtures::dsl::*;
    let conn = database::connection()?;
    let mut inserted_count = 0;
    for new_fixture in incoming.into_iter() {
      diesel::insert_or_ignore_into(fixtures)
        .values(NewFixture::from(new_fixture))
        .execute(&conn)?;
      inserted_count = inserted_count + 1;
    }
    Ok(inserted_count)
  }

  /// TODO: I'm unsure if we will ever need this?!
  ///       if so, it would override hash-value for sure
  pub fn _update(id_to_update: i32, fixture_updates: FixtureDTO) -> Result<Self, ApiError> {
    use crate::schema::fixtures::dsl::*;
    let conn = database::connection()?;
    let pred = id.eq(id_to_update);
    diesel::update(fixtures)
      .filter(&pred)
      .set(NewFixture::from(fixture_updates))
      .execute(&conn)?;
    let res = fixtures.filter(&pred).first(&conn)?;
    Ok(res)
  }

  pub fn _delete(id_to_delete: i32) -> Result<usize, ApiError> {
    use crate::schema::fixtures::dsl::*;
    let conn = database::connection()?;
    let res = diesel::delete(fixtures.filter(id.eq(id_to_delete))).execute(&conn)?;
    Ok(res)
  }
}

impl From<Fixture> for PublicFixture {
  fn from(fixture: Fixture) -> Self {
    PublicFixture {
      name: fixture.name,
      link: fixture.link,
      team1: fixture.team1,
      team2: fixture.team2,
      start_time: fixture.start_time,
      rating: fixture.rating,
      meta: fixture.meta,
      analytics: fixture.analytics,
      top_tier: fixture.top_tier,
      hash: fixture.hash,
      created_at: fixture.created_at,
      updated_at: fixture.updated_at,
    }
  }
}

impl From<Fixture> for FixtureDTO {
  fn from(fixture: Fixture) -> Self {
    FixtureDTO {
      id: None,
      name: Some(fixture.name),
      link: Some(fixture.link),
      team1: Some(fixture.team1),
      team2: Some(fixture.team2),
      start_time: Some(fixture.start_time),
      rating: Some(fixture.rating),
      meta: Some(fixture.meta),
      analytics: Some(fixture.analytics),
      top_tier: Some(fixture.top_tier),
      hash: Some(fixture.hash),
      created_at: Some(fixture.created_at),
      updated_at: Some(fixture.updated_at.unwrap()),
    }
  }
}

/// "Enrich" provided dto values with a hashid (uuid for now) and return an insertable NewFixture object
impl From<FixtureDTO> for NewFixture {
  fn from(new_fixture: FixtureDTO) -> Self {
    let new_uuid = Uuid::new_v4().to_string();

    NewFixture {
      name: new_fixture.name.unwrap_or_default(),
      link: new_fixture.link.unwrap_or_default(),
      team1: new_fixture.team1.unwrap_or_default(),
      team2: new_fixture.team2.unwrap_or_default(),
      start_time: new_fixture
        .start_time
        .unwrap_or_else(|| Utc::now().naive_utc()),
      rating: new_fixture.rating.unwrap_or_default(),
      meta: new_fixture.meta.unwrap_or_default(),
      analytics: new_fixture.analytics.unwrap_or_default(),
      top_tier: new_fixture.top_tier.unwrap_or_default(),
      hash: new_uuid,
    }
  }
}
