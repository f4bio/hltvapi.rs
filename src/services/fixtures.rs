extern crate reqwest;
extern crate scraper;

use std::fs::File;
use std::io::Write;
use std::path::Path;
use std::str::FromStr;
use std::time::Instant;

use chrono::{DateTime, Duration, NaiveDateTime, TimeZone, Utc};
use icalendar::{Calendar, Component};
use scraper::{ElementRef, Html, Selector};
use tracing::{debug, info};

use crate::constants::{
  BASE_URL, FILE_STORAGE_LOCATION, FIXTURES_CALENDAR_FILENAME, FIXTURES_URL, FIXTURES_URL_TOP_TIER,
};
use crate::models::fixture::{Fixture, FixtureDTO};
use crate::services::do_fetch;

fn build_fixture(raw_html: ElementRef) -> Option<FixtureDTO> {
  // # ------ IS EMPTY
  let is_empty_selector = Selector::parse("div.matchInfoEmpty").unwrap();
  let is_empty_element = raw_html.select(&is_empty_selector).next();
  if is_empty_element.is_some() {
    return None;
  };

  // # ------ NAME
  let name_selector = Selector::parse("div.matchEventName").unwrap();
  let name_element = raw_html.select(&name_selector).next().unwrap();
  let name = name_element.inner_html();
  // debug!("raw_html#name: {}", name);

  // # ------ TEAMS
  let teams_selector = Selector::parse("div.matchTeamName").unwrap();
  let mut teams_iter = raw_html.select(&teams_selector);
  let team1_element = teams_iter.next();
  let team1 = match team1_element {
    Some(t) => t.inner_html(),
    None => String::from("TBD"),
  };
  // debug!("team1: {}", team1);
  let team2_element = teams_iter.next();
  let team2 = match team2_element {
    Some(t) => t.inner_html(),
    None => String::from("TBD"),
  };
  // debug!("team2: {}", team2);

  // # ------ TIME
  let time_selector = Selector::parse("div.matchTime").unwrap();
  let time_element = raw_html.select(&time_selector).next();
  let time_string = time_element.unwrap().inner_html();
  let start_time_obj: DateTime<Utc> = if time_string == "LIVE" {
    Utc::now()
  } else {
    let _ts = time_element.unwrap().value().attr("data-unix").unwrap();
    let _millis = i64::from_str(_ts).unwrap();
    Utc.timestamp_millis(_millis)
  };
  let start_time: NaiveDateTime = start_time_obj.naive_utc();
  // debug!("start_time: {}", start_time);

  // # ------ RATING
  let rating_selector = Selector::parse("div.matchRating").unwrap();
  let rating_element = raw_html.select(&rating_selector).next().unwrap();
  let rating_stars_selector = Selector::parse("i.faded").unwrap();
  let rating = (5 - rating_element.select(&rating_stars_selector).count()) as i32;

  // # ------ META
  let meta_selector = Selector::parse("div.matchMeta").unwrap();
  let meta_element = raw_html.select(&meta_selector).next();
  let meta = meta_element?.inner_html();
  // debug!("raw_html#meta: {}", meta.clone().unwrap_or(String::from("> No meta found <")));

  // # ------ LINK
  let link_selector = Selector::parse("a.match").unwrap();
  let link_element = raw_html.select(&link_selector).next().unwrap();

  let link = String::from(link_element.value().attr("href").unwrap());
  // debug!("raw_html#link: {}", link);

  // # ------ ANALYTICS
  let analytics_selector = Selector::parse("a.matchAnalytics").unwrap();
  let analytics_element = raw_html.select(&analytics_selector).next();
  let analytics = String::from(analytics_element?.value().attr("href").unwrap());

  Some(FixtureDTO {
    id: None,
    name: Some(name),
    link: Some(link),
    team1: Some(team1),
    team2: Some(team2),
    start_time: Some(start_time),
    rating: Some(rating),
    meta: Some(meta),
    analytics: Some(analytics),
    top_tier: Some(false),
    hash: None,
    created_at: None,
    updated_at: None,
  })
}

#[test]
fn test_build_fixture() {
  use std::fs;

  let path = Path::new("tests").join("scraped.fixtures_toptier.example.html");
  let content = fs::read_to_string(path).unwrap();

  let fragment = Html::parse_document(&content);
  let current_fixtures = Selector::parse("div.liveMatch,div.upcomingMatch").unwrap();

  for fix in fragment.select(&current_fixtures) {
    let new_fixture = build_fixture(fix);
    // insta::assert_yaml_snapshot!(new_fixture);
    assert_eq!(
      new_fixture.unwrap().name.unwrap(),
      "ESL Pro League Season 14"
    );

    break;
  }
}

fn update_fixture_calendar() {
  info!("updating fixtures calendar...");

  let mut fixture_calendar = Calendar::new();
  let all_fixtures = Fixture::all().unwrap();

  for fix in all_fixtures {
    let summary = format!(
      "{} - {} vs. {} ({})",
      fix.name, fix.team1, fix.team2, fix.meta
    );
    let mut rating_stars = String::with_capacity(5);
    for cnt in 0..5 {
      if cnt >= fix.rating {
        rating_stars = rating_stars + "☆";
      } else {
        rating_stars = rating_stars + "★";
      }
    }
    let link = format!("{}{}", BASE_URL, fix.link);
    let description = format!(
      "rating: {} ({}/5)\nlink: {}\nanalytics: {}\n",
      rating_stars, fix.rating, link, fix.analytics
    );
    let starts = fix.start_time;
    let ends = fix.start_time + Duration::hours(2);

    let event = icalendar::Event::new()
      .uid(fix.hash.as_str())
      .summary(summary.as_str())
      .description(description.as_str())
      .starts(starts)
      .ends(ends)
      .done();
    fixture_calendar.push(event);
  }
  let cal = fixture_calendar.to_string();

  let path = Path::new(FILE_STORAGE_LOCATION).join(FIXTURES_CALENDAR_FILENAME);
  let mut file = File::create(path.clone()).unwrap();
  file.write_all(cal.as_ref()).unwrap();

  info!("fixtures calendar stored at: {}!", path.to_str().unwrap());
}

async fn do_scrape(url: &str) -> Vec<FixtureDTO> {
  let mut fixtures: Vec<FixtureDTO> = Vec::new();
  let body: String = do_fetch(url).await.unwrap();
  let fragment = Html::parse_document(&body);
  let current_fixtures = Selector::parse("div.upcomingMatch").unwrap();
  for fix in fragment.select(&current_fixtures) {
    let new_fixture = build_fixture(fix);
    if new_fixture.is_some() {
      fixtures.push(new_fixture.unwrap());
    }
  }
  return fixtures;
}

pub async fn scrape_fixtures() {
  info!("scraping fixtures...");

  let mut fixtures: Vec<FixtureDTO> = Vec::new();
  fixtures.append(&mut do_scrape(FIXTURES_URL).await);
  fixtures.append(&mut do_scrape(FIXTURES_URL_TOP_TIER).await);

  debug!("adding matches to database...");
  let inserted_fixture = Fixture::create_many(fixtures);
  if inserted_fixture.is_err() {
    info!("fixture already exists, skipping...",);
  }
  info!(
    "inserted '{}' new fixtures",
    inserted_fixture.as_ref().unwrap()
  );

  debug!("adding matches to calendar...");
  let start = Instant::now();
  update_fixture_calendar();
  let duration = start.elapsed();
  debug!(
    "> Time elapsed in update_fixture_calendar() is: {:?}",
    duration
  );

  info!("fixtures scraped!");
}
