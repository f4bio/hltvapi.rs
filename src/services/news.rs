extern crate reqwest;
extern crate scraper;

use std::fs::File;
use std::io::Write;
use std::path::Path;
use std::str::FromStr;
use std::thread;

use chrono::{NaiveDateTime, TimeZone, Utc};
use itconfig::get_env_or_default;
use rss::extension::syndication::SyndicationExtension;
use rss::{Channel, ChannelBuilder, Enclosure, EnclosureBuilder, Guid, Image, ImageBuilder, Item};
use scraper::{ElementRef, Html, Selector};
use tracing::{debug, error, info};

use crate::constants::{BASE_URL, FILE_STORAGE_LOCATION, NEWS_FEED_FILENAME, NEWS_URL};
use crate::models::news::{News, NewsDTO};
use crate::services::do_fetch;

fn build_news(html_content: String) -> Option<NewsDTO> {
  // debug!("build_news: {:?}", html_content);

  let fragment = Html::parse_document(&html_content);
  let news_item_selector = Selector::parse("article.newsitem").unwrap();
  let news_item_element = fragment.select(&news_item_selector).next();
  if news_item_element.is_none() {
    error!(
      "selector 'article.newsitem' not found in html '{:?}'...",
      html_content
    );
    return None;
  };
  let raw_html: ElementRef = news_item_element.unwrap();

  // # ------ Title
  let title_selector = Selector::parse("h1.headline").unwrap();
  let title = match raw_html.select(&title_selector).next() {
    Some(tit) => tit.inner_html(),
    None => String::from("<TITLE>"),
  };
  debug!("build_news#title: {}", title);

  // # ------ LINK
  // og:url
  let link_selector = Selector::parse(r#"meta[property="og:url"]"#).unwrap();
  let link = match fragment.select(&link_selector).next() {
    Some(le) => {
      String::from(le.value().attr("content").unwrap()).replace("https://www.hltv.org", "")
    }
    None => String::from("<LINK>"),
  };
  debug!("build_news#link: {}", link);

  // # ------ PUB_DATE
  // let pub_date: NaiveDateTime = start_time_obj.naive_utc();
  let pub_date_selector = Selector::parse("div.date").unwrap();
  let pub_date_element = raw_html.select(&pub_date_selector).next().unwrap();
  let pub_date_ts = pub_date_element.value().attr("data-unix").unwrap();
  let pub_date_millis = i64::from_str(pub_date_ts).unwrap();
  let pub_date_obj = Utc.timestamp_millis(pub_date_millis);
  let pub_date: NaiveDateTime = pub_date_obj.naive_utc();
  debug!("build_news#pub_date: {}", pub_date);

  // # ------ AUTHOR
  let author_selector = Selector::parse("a.authorName>span").unwrap();
  let author = match fragment.select(&author_selector).next() {
    Some(ae) => ae.inner_html(),
    None => String::from("<AUTHOR>"),
  };

  // # ------ IMAGE
  let image_selector = Selector::parse("div.image-con>img.image").unwrap();
  let image = match raw_html.select(&image_selector).next() {
    Some(img) => String::from(img.value().attr("src").unwrap()),
    None => String::from("<IMAGE>"),
  };

  // # ------ IMAGETEXT
  let image_text_selector = Selector::parse("div.imagetext").unwrap();
  let image_text = match raw_html.select(&image_text_selector).next() {
    Some(img_txt) => img_txt.html(),
    None => String::from("<IMAGE TEXT>"),
  };

  // # ------ CONTENT
  let content_selector = Selector::parse("div.newstext-con>p.news-block").unwrap();
  let mut content_collector = String::new();
  for content_block in raw_html.select(&content_selector) {
    debug!("content block: {}", content_block.html().as_str());
    content_collector.push_str(content_block.html().as_str());
  }

  // # ------ DESCRIPTION
  let description_selector = Selector::parse("p.headertext").unwrap();
  let description = match raw_html.select(&description_selector).next() {
    Some(desc) => desc.inner_html(),
    None => String::from("<DESCRIPTION>"),
  };

  Some(NewsDTO {
    id: None,
    title: Some(title),
    link: Some(link),
    pub_date: Some(pub_date),
    image: Some(image),
    image_text: Some(image_text),
    content: Some(content_collector),
    description: Some(description),
    author: Some(author),
    hash: None,
    created_at: None,
    updated_at: None,
  })
}

#[test]
fn test_build_news() {
  use std::fs;

  let path = Path::new("tests").join("scraped.news_item.example.html");
  let content = fs::read_to_string(path).unwrap();

  let item = build_news(content);

  // insta::assert_yaml_snapshot!(item.unwrap().title);
  assert_eq!(
    item.unwrap().title.unwrap(),
    "EliGE shines as Liquid hammer NIP in ESL Pro League opener"
  );
}

fn update_news_feed() {
  info!("updating news feed...");
  let app_base_url: String = get_env_or_default("PUBLIC_BASE_URL", "https://hltvapi.f4b.io");

  let image: Image = ImageBuilder::default()
    .url("https://hltvapi.f4b.io/logo.png")
    .title("hltvapi logo")
    .build();
  let mut channel: Channel = ChannelBuilder::default()
    .title("Unofficial hltv.org News")
    .link(&app_base_url)
    .description("An unofficial hltv.org RSS news-feed.")
    .image(image)
    .build();
  let all_news = News::all().unwrap();
  let mut news_items: Vec<Item> = vec![];

  for ne in all_news {
    let guid: Guid = Guid {
      value: ne.hash,
      permalink: true,
    };

    let enclosure: Enclosure = EnclosureBuilder::default()
      .mime_type(mime::IMAGE_JPEG.to_string())
      .url(ne.image.unwrap())
      .build();

    // push items into channel
    news_items.push(Item {
      title: Some(ne.title),
      link: Some(format!("{}{}", &app_base_url, ne.link)),
      description: ne.description,
      author: ne.author,
      categories: vec![],
      comments: None,
      guid: Some(guid),
      enclosure: Some(enclosure),
      pub_date: Some(ne.pub_date.to_string()),
      source: None,
      content: ne.content,
      extensions: Default::default(),
      itunes_ext: None,
      dublin_core_ext: None,
    });
  }
  // info!("news items: {:?}", news_items);
  channel.set_syndication_ext(SyndicationExtension::default());
  channel.set_last_build_date(Utc::now().to_rfc2822());
  channel.set_items(news_items);
  let news_feed_rss = channel.to_string();

  let news_feed_rss_path = Path::new(FILE_STORAGE_LOCATION).join(NEWS_FEED_FILENAME);
  let mut news_feed_rss_file = File::create(news_feed_rss_path.clone()).unwrap();
  news_feed_rss_file
    .write_all(news_feed_rss.as_ref())
    .unwrap();

  // let mut opml = OPML::default();
  // let app_url: String = get_env_or_default("APP_SERVER_URL", "http://localhost:1337");
  // let opml_url = format!("{}/news/feed.xml", app_url);
  //
  // opml.head = Some(Head {
  //   title: Some("Rust Feeds".to_string()),
  //   ..Head::default()
  // });
  // opml.add_feed("hltv.org News", opml_url.as_str());
  //
  // let xml = opml.to_string().unwrap();

  info!(
    "news feed rss stored at: {}!",
    news_feed_rss_path.to_str().unwrap()
  );
}

pub async fn scrape_news() {
  info!("scraping base news...");

  let body: String = do_fetch(NEWS_URL).await.unwrap();
  let fragment = Html::parse_document(&body);
  let mut url_list: Vec<String> = Vec::with_capacity(5);

  let news_lines = Selector::parse("a.newsline").unwrap();
  for news_line in fragment.select(&news_lines) {
    let link = String::from(news_line.value().attr("href").unwrap());
    url_list.push(link);
    if url_list.len() >= 5 {
      break;
    }
  }

  for url_path in url_list {
    thread::spawn(move || {
      let resp = reqwest::blocking::get(format!("{}{}", BASE_URL, url_path));
      let body = resp.unwrap().text().unwrap();

      let news_item = build_news(body);
      let inserted_news = News::create(news_item.unwrap());
      if inserted_news.is_err() {
        info!("news already exists, skipping...");
      }
    });
  }

  debug!("adding news to feed...");
  update_news_feed();

  info!("news scraped!");
}
