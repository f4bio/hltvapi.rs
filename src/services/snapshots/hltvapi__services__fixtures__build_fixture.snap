---
source: src/services/fixtures.rs
expression: new_fixture

---
id: ~
name: ESL Pro League Season 14
link: /matches/2350366/furia-vs-teamone-esl-pro-league-season-14
team1: FURIA
team2: TeamOne
start_time: "2021-09-02T11:30:00"
rating: 1
meta: bo3
analytics: /betting/analytics/2350366/furia-vs-teamone-esl-pro-league-season-14
top_tier: true
hash: ~
created_at: ~
updated_at: ~
